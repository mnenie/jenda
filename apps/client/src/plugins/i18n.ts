import { createI18n } from 'vue-i18n'
import { enZod, ruZod, zhZod } from '@/shared/libs/vee-validate/rules'
import { customPluralRule } from '@/shared/libs/i18n/plurals'
import { getCurrentLocale } from '@/shared/libs/i18n/utils/getLocale'

interface MessageSchema {
  [key: string]: any
}

const ru_ru = import.meta.glob(
  '@/shared/libs/i18n/locales/**/ru-RU.ts',
  { eager: true },
)
const en_us = import.meta.glob(
  '@/shared/libs/i18n/locales/**/en-US.ts',
  { eager: true },
)
const zh_cn = import.meta.glob(
  '@/shared/libs/i18n/locales/**/zh-CN.ts',
  { eager: true },
)

function loadLang(modules: Record<string, any>) {
  const messages: { [key: string]: string } = {}

  Object.keys(modules).forEach((module) => {
    Object.assign(messages, { ...modules[module].default })
  })
  return messages
}

const messages = {
  'en-US': {
    ...loadLang(en_us),
    errors: {
      ...enZod,
    },
  },
  'ru-RU': {
    ...loadLang(ru_ru),
    errors: {
      ...ruZod,
    },
  },
  'zh-CN': {
    ...loadLang(zh_cn),
    errors: {
      ...zhZod,
    },
  },
}

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  pluralRules: {
    'ru-RU': {
      // @ts-expect-error i18n shema
      customPluralRule,
    },
  },
  messages,
})

export default i18n
