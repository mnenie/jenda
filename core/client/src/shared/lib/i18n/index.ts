import { useCookies } from '@vueuse/integrations/useCookies'
import { computed } from 'vue'
import { createI18n } from 'vue-i18n'
import enLocale from './locales/en-US'
import ruLocale from './locales/ru-RU'
import zhLocale from './locales/zh-CN'

type MessageSchema = typeof enLocale

const messages = {
  'en-US': {
    ...enLocale,
  },
  'ru-RU': {
    ...ruLocale,
  },
  'zh-CN': {
    ...zhLocale,
  },
}
const cookies = useCookies()

const getCurrentLocale = computed(() => {
  const cookieLanguage = cookies.get('i18n')
  if (cookieLanguage) {
    return cookieLanguage
  }
  if (navigator.language.split('-')[0] === 'ru') {
    return 'ru-RU'
  }
  return 'en-US'
})

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  messages,
})

export default i18n
