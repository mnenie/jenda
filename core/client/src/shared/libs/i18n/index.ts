import { computed } from 'vue'
import { createI18n } from 'vue-i18n'

import en from '../vee-validate/rules/en'
import ru from '../vee-validate/rules/ru'
import zh from '../vee-validate/rules/zh'
import enLocale from './locales/en-US'
import ruLocale from './locales/ru-RU'
import zhLocale from './locales/zh-CN'

export type MessageSchema = typeof enLocale

const messages = {
  'en-US': {
    ...enLocale,
    errors: {
      ...en,
    },
  },
  'ru-RU': {
    ...ruLocale,
    errors: {
      ...ru,
    },
  },
  'zh-CN': {
    ...zhLocale,
    errors: {
      ...zh,
    },
  },
}

const getCurrentLocale = computed(() => {
  const storageLanguage = localStorage.getItem('i18n')
  if (storageLanguage) {
    return storageLanguage
  }
  if (navigator.language.split('-')[0] === 'ru') {
    return 'ru-RU'
  }
  return 'en-US'
})

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  messages,
})

export default i18n
