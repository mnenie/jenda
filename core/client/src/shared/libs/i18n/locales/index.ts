export { default as enLocale } from './en-US'
export { default as ruLocale } from './ru-RU'
export { default as zhLocale } from './zh-CN'
