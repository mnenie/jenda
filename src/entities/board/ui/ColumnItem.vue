<script setup lang="ts">
import { useDark } from '@vueuse/core';
import { Ellipsis, Plus } from 'lucide-vue-next';
import type { Column } from '../model';
import { useI18n } from 'vue-i18n';

defineOptions({
  inheritAttrs: false
});

defineProps<{
  column: Column;
}>();

const isDark = useDark();
const { t } = useI18n();

defineSlots<{
  content: () => any;
  default: () => any;
}>();
</script>

<template>
  <div :class="$style.column">
    <div class="text-sm" :class="$style.top_part" v-bind="$attrs">
      <div :class="$style.name">
        <slot />
        <span>{{ column.cards?.length }}</span>
      </div>
      <Ellipsis :size="16" :color="isDark ? 'var(--zinc-300)' : 'var(--zinc-400)'" />
    </div>
    <div v-if="column.cards" :class="$style.content">
      <slot name="content" />
    </div>
    <div :class="$style.bottom">
      <div class="text-sm" :class="$style.add">
        <Plus :size="16" :color="isDark ? 'var(--zinc-300)' : 'var(--zinc-500)'" />
        {{ t('kanban.cards.add') }}
      </div>
    </div>
  </div>
</template>

<style module lang="scss">
.column {
  position: relative;
  justify-content: space-between;
  min-width: 260px;
  width: 260px;
  max-height: 100%;
  height: fit-content;
  overflow-y: auto;
  border-radius: 12px;
  background-color: white;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
  vertical-align: top;
  white-space: normal;
  scroll-margin: 8px;

  .top_part {
    position: sticky;
    top: 0;
    left: 0;
    z-index: 999;
    padding: 0.8rem;
    border-radius: 12px 12px 0 0;
    background-color: #f9f9f9;
    cursor: grab;
    display: flex;
    justify-content: space-between;
    align-items: center;
    color: var(--zinc-500);
    font-weight: 500;
  }

  .content {
    padding: 6px;

    & > div {
      display: flex;
      flex-direction: column;
      gap: 6px;
    }
  }

  .name {
    display: flex;
    align-items: center;
    gap: 10px;
    width: 100%;

    .indicator {
      width: 5px;
      height: 5px;
      border-radius: 50%;
    }
  }

  .bottom {
    position: sticky;
    bottom: 0;
    left: 0;
    padding: 0.8rem;
    border-radius: 0px 0px 12px 12px;
    background-color: #f9f9f9;
    display: flex;
    justify-content: space-between;
    align-items: center;
    color: var(--zinc-400);
    font-weight: 500;

    .add {
      display: flex;
      align-items: center;
      gap: 6px;
      cursor: pointer;
    }
  }
}

:global(html.dark) {
  .column {
    background-color: var(--zinc-700);
    .top_part,
    .bottom {
      background-color: rgba(33, 33, 33, 0.6);
      color: var(--zinc-200);
    }
  }
}
</style>
