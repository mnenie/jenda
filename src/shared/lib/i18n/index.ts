import { createI18n } from 'vue-i18n';
import { useCookies } from '@vueuse/integrations/useCookies';
import enLocale from './locales/en-US';
import ruLocale from './locales/ru-RU';
import zhLocale from './locales/zh-CN';

type MessageSchema = typeof enLocale;

const messages = {
  'en-US': {
    ...enLocale
  },
  'ru-RU': {
    ...ruLocale
  },
  'zh-CN': {
    ...zhLocale
  }
};
const cookies = useCookies();

const getCurrentLocale = () => {
  const cookieLanguage = cookies.get('i18n');
  if (cookieLanguage) {
    return cookieLanguage;
  }

  return 'en-US';
};

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale(),
  messages: messages
});

export default i18n;
