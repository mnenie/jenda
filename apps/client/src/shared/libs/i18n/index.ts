import { createI18n, type I18n } from 'vue-i18n'
import { enZod, ruZod, zhZod } from '../vee-validate/rules'
import { enLocale, ruLocale, zhLocale } from './locales'
import { customPluralRule } from './plurals'
import useDayJsLocalize from './dayjs'
import { getCurrentLocale } from './utils/getLocale'

export type MessageSchema = typeof enLocale

const messages = {
  'en-US': {
    ...enLocale,
    errors: {
      ...enZod,
    },
  },
  'ru-RU': {
    ...ruLocale,
    errors: {
      ...ruZod,
    },
  },
  'zh-CN': {
    ...zhLocale,
    errors: {
      ...zhZod,
    },
  },
}

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  pluralRules: {
    'ru-RU': {
      // @ts-expect-error i18n shema
      customPluralRule,
    },
  },
  messages,
})

useDayJsLocalize(i18n as I18n)

export default i18n
