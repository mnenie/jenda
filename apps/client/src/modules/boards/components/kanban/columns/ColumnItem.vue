<script setup lang="ts" generic="T extends Column">
import { computed } from 'vue'
import { Icon } from '@iconify/vue'
import ColumnMenu from './ColumnMenu.vue'
import type { Column } from '../../../types'
import { UiBadge, UiButton } from '@/shared/ui'

defineProps<{
  column: T
}>()

const colors = ['#32a852', '#c2c734', '#a934c9', '#c93434', '#309dc2']

// пофиксить с бэком
const getRandomColor = computed(() => {
  return colors[Math.floor(Math.random() * colors.length)]
})
</script>

<template>
  <div class="rounded-md h-fit flex flex-col justify-between gap-2 max-w-280px w-full p-1 bg-neutral-50 dark:bg-neutral-800 border border-neutral-200 dark:border-neutral-700">
    <div class="flex items-start flex-col gap-1 w-full">
      <div class="flex items-center justify-between p-2 h-7 w-full">
        <div class="flex items-center gap-1">
          <div
            class="w-4 h-4 rounded-full border-1.5"
            :style="{ borderColor: getRandomColor, backgroundColor: `${getRandomColor}20` }"
          />
          <span class="text-default ml-0.5 fw500 text-neutral-800 dark:text-neutral-200">
            {{ column.title }}
          </span>
          <span class="text-neutral-500 dark:text-neutral-400">{{ column.cards?.length }}</span>
          <UiBadge
            variant="outline"
            class="ml-1.5 text-neutral-500 dark:text-neutral-400 rounded-xl"
          >
            Estimate: {{ column.estimate }}
          </UiBadge>
        </div>
        <ColumnMenu />
      </div>
      <span v-if="column.description" class="text-default ml-2 fw500 text-neutral-400 dark:text-neutral-500">
        {{ column.description }}
      </span>
    </div>
    <slot v-if="column.cards?.length" name="column-content" :column :cards="column.cards" />
    <div>
      <UiButton variant="ghost" size="sm" class="w-full justify-start h-8">
        <Icon icon="lucide:plus" class="min-w-4.4 min-h-4.4 text-neutral-500 dark:text-neutral-400" />
        <span class="text-default fw500 text-neutral-600 dark:text-neutral-400">
          {{ $t('kanban.column.tasks.new') }}
        </span>
      </UiButton>
    </div>
  </div>
</template>
