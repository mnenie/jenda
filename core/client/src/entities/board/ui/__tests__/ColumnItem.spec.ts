import i18n from "@/shared/lib/i18n";
import { mount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";
import ColumnItem from "../ColumnItem.vue";
import "@/shared/lib/vitest-utils/cookiesI18n-mock";

describe("tests for ColumnItem.vue", () => {
  const wrapper = mount(ColumnItem, {
    global: {
      plugins: [i18n],
      mocks: {
        t: (key: string) => {
          const translations: Record<string, string> = {
            "kanban.cards.add": "Add card (test)",
          };
          return translations[key];
        },
      },
    },
    props: {
      column: {
        _id: "0",
        title: "test",
        cards: [],
      },
    },
    slots: {
      default: "test name of column",
      content: "test content",
    },
  });

  it("should render correctly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should correctly works with i18n", () => {
    const date = wrapper.find(".add");
    expect(date.text()).toContain("Add card (test)");
  });
});
