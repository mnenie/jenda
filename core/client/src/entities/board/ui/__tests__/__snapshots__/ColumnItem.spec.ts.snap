// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tests for ColumnItem.vue > should render correctly 1`] = `
"<div class="column">
  <div class="text-sm top_part">
    <div class="name">test name of column<span>0</span></div>
    <div i-lucide-ellipsis="" text-base=""></div>
  </div>
  <div class="content">test content</div>
  <div class="bottom">
    <div class="text-sm add">
      <div i-hugeicons-plus-sign="" text-base=""></div> Add card (test)
    </div>
  </div>
</div>"
`;
